keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1158 @@
+  relevant trust anchors using a TrustAnchorSets object.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregrated API server
+  backends using that use TrustAnchorSets instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors using TrustAnchorSets.
+  (This is a natural extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped TrustAnchorSets resource,
+initially in the certificates/v1alpha1 API group.  The TrustAnchorSets object
+can be thought of as a specialized configmap tailored to the X.509 trust anchor
+use case.  Introducing a dedicated type allows us to attach different RBAC
+policies to TrustAnchorSets, which will typically be wide-open for reading, but
+locked-down for writing.
+
+Each TrustAnchorSets object is a container for multiple named trust anchor sets.
+Most use cases will simply use a single trust anchor set named `default`, but
+mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
+
+TrustAnchorSets are scoped to a particular signer using a naming convention that
+translates the signer name into a valid Kubernetes resource name.  For example,
+the signer named `example.com/my-signer` will conventionally export any
+necessary trust anchor sets using a TrustAnchorSets object named
+`example.com-slash-my--signer`.
+
+To enable consumption by workloads, the new `trustAnchorSets` Kubelet projected
+volume source allows projecting a trust anchor set from a particular signer into
+the container filesystem, with the contents of the projected files updating as
+the corresponding TrustAnchorSets are updated.
+
+### User Stories
+
+#### Trust Anchor Distribution for Private CAs
+
+In my cluster, I operate a certificates/v1 signer implementation called
+`example.com/server-tls`.  This is backed by a private CA hierarchy.  By some
+unspecified mechanism, I issue certificates from this CA hierarchy to a server
+workload (a pod named `server/server`).  My client workload (a pod named
+`client/client`) needs access to the trust anchors for the CA in order to
+connect to the server.  The client should be able to continually establish
+connections to the server without downtime, even as I rotate the trust anchors
+that back the CA.
+
+To do this, I update the controller that implements `example.com/server-tls` to
+create a TrustAnchorSets object like this one:
+```yaml
+apiVersion: v1alpha1
+kind: TrustAnchorSets
+metadata:
+  name: example.com--server-tls
+status:
+  sets:
+  - name: default
+    pemTrustAnchors:
+    - "<... PEM DATA ...>"
+```
```

Comment:
According to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object, this isn't an object as it has `status` but no `spec`.

I would rename `status` to `anchors` and then omit `anchors.sets`, moving the array to be a child of `anchors`.

I recommend asking SIG Architecture for an opinion on the precedent this might set.

Reply:
There is also `.status.conditions`, which the controller uses to indicate whether the contents of the object are ready to use.  They might be empty or stale during startup of the signer's controller, or during an outage of the backing CA.

I don't think this case is covered in the conventions doc.  ClusterTrustAnchorConfig is an object with no user intent, only system-generated state.  They mention the mirrored case (object with only user intent, no system-generated state).

I'll ask SIG Architecture for feedback.

Reply:
1. We could put the signer name in the `.spec` and make it an optional field.
1. We could update the API conventions doc to cover this case.

I'd pick the first option, TBH. Even if the name is derived from the spec, it gives us a place to put the signer name.

Reply:
I like putting the signer name in, but we will need to add some sort of admission control to verify that one signer can't spoof another.

Reply:
Following up: there is now a `.spec.signerName` (optional) field, with admission control to ensure that it can't be spoofed.